Show the contract receiving token allowances on the allowance screen

blackdevelopa opened the pull request:
<!--
Thanks for your contribution!

Please ensure that any applicable requirements below are satisfied before submitting this pull request. This will help ensure a quick and efficient review cycle.
-->

**Description**

Contract address naming

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [x] Any added code is fully documented

**Issue**

Resolves #2723

blackdevelopa committed:
show contract receiving token and allow to add a nickname

blackdevelopa committed:
updates on show contract receiving details

blackdevelopa committed:
fixed copy feedback

blackdevelopa committed:
design updates

sethkfman commented on app/components/UI/ApproveTransactionReview/index.js:
```diff
@@ -552,11 +607,37 @@ class ApproveTransactionReview extends PureComponent {
 						)}`}
 					</Text>
+					<View style={styles.contactWrapper}>
+						<Text>Contract: </Text>
```
Move this string to locals so it can be translated.

sethkfman commented on app/components/UI/ApproveTransactionReview/index.js:
```diff
@@ -552,11 +607,37 @@ class ApproveTransactionReview extends PureComponent {
+						</TouchableOpacity>
+					</View>
+					<Text style={styles.nickname} onPress={this.toggleDisplay}>
+						{this.props.nicknameExists ? 'Edit' : 'Add'} nickname
```
Move strings to locales for translation.

sethkfman reviewed (commented):
Minor updates required to use locales

blackdevelopa committed:
added text to translation

cortisiko reviewed (changes requested):
**Issue 1**
I get this warning:  `ERROR  Warning: Failed prop type: Invalid prop protectWalletModalVisible of type function supplied to ApproveTransactionReview, expected boolean.`

[When I attempted to approve a token
](http://recordit.co/BOdVyp8Ujp)

to reproduce:
go [here](https://metamask.app.link/dapp/metamask.github.io/test-dapp/) while on a test net (because you do not want to use real funds)  
connect your wallet 
scroll down on the page and tap on create token
after your txn was confirmed tap on “approve token”
notice the confirmation modal appears with the error


**Issue 2**
after saving the contract address to my contacts I get: `Cannot update during an existing state transition (such as within render). Render methods should be a pure function of props and state`

 see [here](http://recordit.co/QF8tDB1kRv )

to reproduce:
go [here](https://metamask.app.link/dapp/metamask.github.io/test-dapp/) while on a test net (because you do not want to use real funds)  
connect your wallet 
scroll down on the page and tap on create token
after your txn was confirmed tap on “approve token”
notice the confirmation modal appears. Now tap on “Add nickname” 
input a nick name then hit save

notice the error appears 

**Issue 3**
On the edit nick name view I am unable to share my address. Tapping on the share icon does nothing.  On the other hand tapping on the address does not copy it to my clipboard. I need to carefully tap on the copy icon in order to copy. Even when i do copy my address i see a toast message “shell was pasted from your clipboard”. Maybe it should be “address was copied to your clipboard” ? 

see [here](http://recordit.co/VL5k3u4aBb )

**Issue 4**
[the contract i added to my contacts should not appear in recents
](http://recordit.co/gT5KQDeQN6 )
to reproduce:
go [here](https://metamask.app.link/dapp/metamask.github.io/test-dapp/) while on a test net (because you do not want to use real funds)  
connect your wallet 
scroll down on the page and tap on create token
after your txn was confirmed tap on “approve token”
notice the confirmation modal appears. Now tap on “Add nickname” 
input a nick name then hit save
return to the wallet view then go through the send flow
notice the contract address appears

blackdevelopa committed:
fixed QA issue 1 & 3

blackdevelopa committed:
QA issue 2 fixed

blackdevelopa commented:
> **Issue 1** I get this warning: `ERROR Warning: Failed prop type: Invalid prop protectWalletModalVisible of type function supplied to ApproveTransactionReview, expected boolean.`
> 
> [When I attempted to approve a token ](http://recordit.co/BOdVyp8Ujp)
> 
> to reproduce: go [here](https://metamask.app.link/dapp/metamask.github.io/test-dapp/) while on a test net (because you do not want to use real funds) connect your wallet scroll down on the page and tap on create token after your txn was confirmed tap on “approve token” notice the confirmation modal appears with the error
> 

Thank you for catching this, QA issue 1 is now fixed.

> **Issue 2** after saving the contract address to my contacts I get: `Cannot update during an existing state transition (such as within render). Render methods should be a pure function of props and state`
> 
> see [here](http://recordit.co/QF8tDB1kRv)
> 
> to reproduce: go [here](https://metamask.app.link/dapp/metamask.github.io/test-dapp/) while on a test net (because you do not want to use real funds) connect your wallet scroll down on the page and tap on create token after your txn was confirmed tap on “approve token” notice the confirmation modal appears. Now tap on “Add nickname” input a nick name then hit save
> 
> notice the error appears
> 

Thank you for catching this, QA issue 2 is now fixed.

> **Issue 3** On the edit nick name view I am unable to share my address. Tapping on the share icon does nothing. On the other hand tapping on the address does not copy it to my clipboard. I need to carefully tap on the copy icon in order to copy. Even when i do copy my address i see a toast message “shell was pasted from your clipboard”. Maybe it should be “address was copied to your clipboard” ?
> 
> see [here](http://recordit.co/VL5k3u4aBb)
> 

Thank you for catching this, QA issue 3 is now fixed. We've got approval from Design on the `share` icon

> **Issue 4** [the contract i added to my contacts should not appear in recents ](http://recordit.co/gT5KQDeQN6) to reproduce: go [here](https://metamask.app.link/dapp/metamask.github.io/test-dapp/) while on a test net (because you do not want to use real funds) connect your wallet scroll down on the page and tap on create token after your txn was confirmed tap on “approve token” notice the confirmation modal appears. Now tap on “Add nickname” input a nick name then hit save return to the wallet view then go through the send flow notice the contract address appears

Thank you for catching this, QA issue 4 might take a little longer as it will require an update on the AddressController file to complete this.

blackdevelopa committed:
added a check

blackdevelopa committed:
fix issue 4

blackdevelopa committed:
design feedback from demo

blackdevelopa committed:
added longpress to contract address when editing one

blackdevelopa committed:
fix bug with displaying empty label, styling transactionreview

sethkfman commented on app/components/Views/ApproveView/Approve/index.js:
```diff
@@ -44,6 +47,10 @@ const styles = StyleSheet.create({
 	},
+	updateNickView: {
+		margin: 0,
+		// backgroundColor: '#fff',
```
remove comment

sethkfman commented on app/components/Views/ApproveView/Approve/index.js:
```diff
@@ -518,7 +538,23 @@ class Approve extends PureComponent {
 
+	checkIfAddressIsSaved = () => {
+		const { addressBook, network, transaction } = this.props;
+		for (const [key, value] of Object.entries(addressBook)) {
```
consider breaking this out to a util function and testing it

sethkfman commented on app/components/UI/ApproveTransactionReview/index.js:
```diff
@@ -415,6 +462,8 @@ class ApproveTransactionReview extends PureComponent {
 			content: 'clipboard-alert',
 			data: { msg: strings('transactions.address_copied_to_clipboard') },
 		});
+		setTimeout(() => this.props.protectWalletModalVisible(), 2000);
```
Can we avoid this setTimeout?

blackdevelopa committed:
update on feedback

cortisiko commented:
@blackdevelopa sending this back your way:


**Issue 1**

I get this error: `[TypeError: undefined is not an object (evaluating '_$$_REQUIRE(_dependencyMap[3], "ethjs-util").stripHexPrefix(address).toLowerCase')]` when:

I go to https://uniswap.exchange/
Connect your wallet
Toggle from Swap to Charts.
Upon doing so I get that error 

See here: http://recordit.co/ZjZJ4L4h7M  
See main: http://recordit.co/flwnspWKMP 

**Maybe updating your branch would fix this?** 


**Issue 2**
I added a nick name to a contract then:

 went to the contacts view 
 tried editing the nickname.
 After saving my edit I am taken back to a blank contacts view. 

See here: http://recordit.co/QKAIXuRhLx 

**Issue 2B**

Similar to the above, If you were to:
add a nick name to a contract 
Go to the wallet view
Tap on send. Verify the contract address does not appear
Now go to the contacts view 
Notice the saved contract address is not visible. 

See here: http://recordit.co/MawUyJKiAh

sethkfman reviewed (commented):
Architecture comments

blackdevelopa committed:
contract address shows in contact list and not send flow

sethkfman reviewed (approved):
LGTM

cortisiko committed:
creating e2e tests

blackdevelopa committed:
fix the delay with updating the sendflow

owencraston commented on app/components/UI/ApproveTransactionReview/AddNickNameHeader/index.tsx:
```diff
@@ -0,0 +1,27 @@
+}
+
+const Header = (props: HeaderProps) => {
+	const { onUpdateContractNickname, nicknameExists, headerWrapperStyle, headerTextStyle, iconStyle } = props;
```
NIT PICK (Not important): you could just destructure these values in the function declaration.
```
const Header = ({onUpdateContractNickname, nicknameExists, headerWrapperStyle, headerTextStyle, iconStyle}: HeaderProps) => {
```

owencraston commented on app/components/UI/ApproveTransactionReview/ShowBlockExplorer/index.tsx:
```diff
@@ -0,0 +1,65 @@
+	},
+});
+
+interface ShowBlockExplorerProps {
```
I can see that this is a very common pattern so we can discuss this with the team but I am really not a fan of putting the logic in the the index.ts/js. I reason for this is because when you are working with many files, all of the files that you open look the same, making it super difficult to navigate. 

I know the reason for this is so you can get default imports but what I prefer is to create the folder `ShowBlockExplorder` and the have the logic inside a file `ShowBlockExplorder.tsx` and add an `index.ts` file in the folder that consist of ...
```
// index.ts
import {ShowBlockExplorder} from './ShowBlockExplorder';

export default ShowBlockExplorder;
```

owencraston commented on app/components/UI/ApproveTransactionReview/ShowBlockExplorer/index.tsx:
```diff
@@ -0,0 +1,65 @@
+	);
+};
+
+export default ShowBlockExplorer;
```
This component can be wrapped in `React.memo(...)`
https://reactjs.org/docs/react-api.html#reactmemo

Pretty much every component should be wrapped in this unless its a component that lives very high in the tree (wraps a lot of views/components)

owencraston commented on app/components/UI/ApproveTransactionReview/index.js:
```diff
@@ -55,7 +59,7 @@ const styles = StyleSheet.create({
 		color: colors.black,
 		lineHeight: 34,
-		marginVertical: 16,
+		marginVertical: 8,
```
Not that we need to do this now but we should definitely create  file that has our constant style baselines. Having each stylesheet inline the values leads to inconsistency and becomes super hard to change if the designers decide we need to overall all of the spacing.

what I envision is ...
a file called `StyleConstants.ts` (or something like that)
```
// StyleConstants.ts`
const baseMarginVertical = 4;
```

and then in your stylesheet you can do ...
`marginVertical: 2*baseMarginVertical,`
another option would be to turn the style units into a function so we can have something like this
```
const baseMarginVertical = 4;
const spaces = (multiplier: number) => baseMarginVertical *  multiplier;
```
and then in the styles we would do...
`marginVertical: spaces(2)`

owencraston replied:
I talked to @Cal-L and he said something like this is in the words on the redesign team.

blackdevelopa replied:
I actually like this idea and I am all for it. We can sync with the rest of the team and adopt this approach.

owencraston commented on app/components/UI/ApproveTransactionReview/index.js:
```diff
@@ -91,6 +95,21 @@ const styles = StyleSheet.create({
+		backgroundColor: colors.blue000,
+		flexDirection: 'row',
+		alignItems: 'center',
+		borderRadius: 40,
```
as you can see keeping track of all these raw numbers can get super tedious and could lead to inconsistent styles. Not to mention it adds an extra burden on the developer to know what the standard values for certain properties are, if these were all constants/functions, we would all just rely on them and the burden would be lifted.

blackdevelopa replied:
On the same page with you on this. We should definitely sync up on this.

owencraston commented on app/components/Views/ApproveView/Approve/index.js:
```diff
@@ -113,6 +119,14 @@ class Approve extends PureComponent {
+		/**
+		 * The current network of the app
+		 */
+		network: PropTypes.string,
```
I'm unfamiliar with this library. why not just use typescript primitive types?

blackdevelopa committed:
before backing seedphrase, the app freezes when address is copied in the approval flow, removed the code

cortisiko committed:
test changes

blackdevelopa committed:
introduced isloading when checking for contract address

blackdevelopa committed:
fixed a bug

cortisiko committed:
final test changes

blackdevelopa commented on app/components/UI/ApproveTransactionReview/AddNickname/index.tsx:
```diff
@@ -0,0 +1,260 @@
+import Header from '../AddNickNameHeader';
+import ShowBlockExplorer from '../ShowBlockExplorer';
+
+const styles = StyleSheet.create({
```
I also agree with you on this, but at this time the team's practice is to have the styles where it is - at the top. We can always review this for future updates.

cortisiko committed:
close send view.

blackdevelopa committed:
broken edge case

cortisiko committed:
move contract nickname into it's own file.

cortisiko reviewed (approved):
🌮 🌮 🌮

cortisiko committed:
fixing test because keyboard issue on CI.

blackdevelopa committed:
Merge branch 'main' into feature/2723-approve-hostname
